File: input_error_MAT/Database/16.plug pmf_household/input_household.m
function data_out_house = input_household(n,error,x)
data_out_house = zeros(n,3);
soc_pmf = readmatrix('pmf_ini_SOC.csv');

%% departure time pdf
mu = 9.97;
sigma = 2.2;

r = normrnd(mu,sigma,[n,1]);
R = round(r); % in-time

idx = find(R<1);
if isempty(idx) == 0
    R(idx) = R(idx) + 24;
end
data_out_house(:,2) = R;

%% arrival time pdf
mu = 17.01;
sigma = 3.2;

r = normrnd(mu,sigma,[n,1]);
R = round(r); % in-time

idx = find(R>24);
if isempty(idx) == 0
    R(idx) = R(idx) - 24;
end
data_out_house(:,1) = R;

%% sampling: initial soc
for t = 1:n
    v = soc_pmf(:,1);
    p = soc_pmf(:,data_out_house(t,1)+1);
    c = cumsum([0,p(:).']);
    c = c/c(end); % make sur the cumulative is 1
    [~,i] = histc(rand(1),c);
    data_out_house(t,3) = v(i);
end

%% data out
writematrix(data_out_house,'16.plug pmf_household/out_plug_time.csv')
%% error input
if error == 1
    data_out_error = dist_error(x,n);
    writematrix(data_out_error,'16.plug pmf_household/out_plug_time_error.csv')
end

end

File: input_error_py/input_household.py
def input_household(n,x,error,office):
    import pandas as pd
    import numpy as np
    import Dist_error
    
    soc_pmf = pd.read_csv("16.plug pmf_household/pmf_ini_SOC.csv",header = None)
    data_out_house = pd.DataFrame(np.zeros([n,3]),columns = ['in','out','init_soc'])
           
    #departure time
    mu = 17.01
    sigma = 3.2
    r = np.random.randn(n)
    R = np.zeros(n)
    for i in range(n):
        R[i] = round(mu + sigma * r[i])
        if R[i] <= 0:
            R[i] = R[i] + 24
        data_out_house['out'][i] = R[i]
        
    #arrival time
    mu = 9.9
    sigma = 2.2
    r = np.random.randn(n)
    R = np.zeros(n)
    for i in range(n):
        R[i] = round(mu + sigma * r[i])
        if R[i] >= 24:
            R[i] = R[i] - 24
        data_out_house['in'][i] = R[i]
    
    #init soc sampling
    c = pd.DataFrame(np.zeros([1,len(soc_pmf)]),columns = None)
    for i in range(n):
        b3 = 0
        v = soc_pmf[0]
        p = soc_pmf[data_out_house['in'][i]]
        cumsum = 0
        for j in range(len(p)):
            cumsum += p[j]
            c[j] = round(cumsum,4)
        c_array = c.values.tolist()
        c_array = c_array[0]
        rand_num = round(float(np.random.rand(1)),4)
        b3 = np.digitize(rand_num,c_array)
        if b3 >= 10:
            b3 = 9
        data_out_house['init_soc'][i] = v[b3]
    
    data_out_house.to_csv("16.plug pmf_household/out_plug_time.csv")
    
    if error == 1:
        data_out_error = Dist_error.Dist_error(n,x,office)
        
